src/docs/binding-classes.md

```diff
@@ -0,0 +1,397 @@
+    [DllImport(/* ... */)]
+    static extern nint SuperGetAge(SwiftSelf<IntPtr> p);
+}
+```
+
+What happens is that the wrapper class in swift defines a vtable type which has a C calling convention callback. It also defines a method to call the
+super class implementation.
+
+What that leaves us with a C# class that when called from C# will call the superclass implementation in Swift. When passed into swift, `getAge` will call the vtable, which will go into C#, get a GCHandle for the actual C# class and call its implementation of `GetAge`, which will then call the Swift superclass implementation.
+
+If a user creates subclasses `YourAge` and overrides `GetAge`, that implementation will be used.
+
+This gets more complicated for generic classes.
+
+For generic classes you don't have a singleton vtable. Instead you need a vtable for every specialization of the class.
+
+In addition, because of issues with the way that C# handles unmanaged callbacks, you can't have an unmanaged callback inside a generic class.
+As a result it's necessary to create an unmanaged vtable and a managed vtable. 
+
+For each unmanaged vtable function of the form `delegate *unmanaged<IntPtr, IntPtr, args..., Result>` there will be a managed delegate of the form
+`Func<IntPtr, args, Result>`. The unmanaged vtable will also have an IntPtr which is a `GCHandle` to the managed table. The unmanaged delegate will include an `IntPtr` as the first argument which is the handle to the managed vtable.
+
+Therefore a receiver in C# will look something like this:
+
+```csharp
+[UnmanagedCallersOnly]
+public static Result SomeUnmanagedReceiver(IntPtr managedVTPtr, IntPtr csHandle, args...)
+{
+    var gch = GCHandle.FromIntPtr(managedVTPtr);
+    if (gch.Target is ManagedVT mvt) {
+        return mvt.SomeManagedReceiver(csHandle, args)
+    }
+}
+```
+
+One thing that this does not do is expose new members to Swift nor does it make protocol compliance visible in all circumstances.
+For example, if I do something like this in C#:
+```csharp
+public class OneAge : YourAge, ISomeSwiftProtocol {
+    public int ExtraProperty { get; set; }
+    public void DoSomethingCool ()
+    {
+        TheBoundNamespace.SomeOtherFunc(this);
+    }
+}
+```
+and had this in Swift:
+```swift
+public func someOtherFunc(y: YourAge) {
+    if let ssp = y as? SomeSwiftProtocol { // as called from C# this will never work since y is YourAgeWrapper
+        ssp.ProtocolAction()
+    }
+}
+```
+
+The way to get this to work 100% in both directions is to have a reverse binding step that exposes the new C# members and protocol compliance in Swift.
+
+It may seem tempting to synthesize the protocol witness table, a type object and a synthetic swift object, but this would require knowing the layout of the protocol witness tables as well as the layout of the actual vtables which we don't and can't know. In addition, both tables are fragile and susceptible to changes in the class and/or protocol(s).
```

Comment:
I mean technically, yes, we could generate it at runtime in the sense that it's just a block of data in memory, but the ordering is unknown. In theory, if you know how big the protocol witness table is, you could line up the address of the entries with methods in the dylib to figure out the ordering. This gets more challenging when the type is generic because really wants to own the creation process of, say, the protocol witness table for a type conforming to a protocol with associated types.

Reply:
As mentioned in another comment, can we work with Apple to document the existing ordering of methods in protocol witness tables in enable-library-evolution scenarios? Then the projection tooling could use those rules to order the members in its generated code.

Apple can't change the algorithm of the vtable ordering in the `enable-library-evolution` scenarios, even for `@frozen` types between compilations (or they have to have some other stable mechanism), so we can use that ordering/mechanism reliably (as non-enable-library-evolution scenarios are out of scope).

Reply:
Enable-library-evolution is designed so that witness tables ordering can change and not break consuming code. Apple lays that out [here](https://www.swift.org/blog/library-evolution/) and are explicit in saying that the layout much be completely opaque:

> When a framework publishes a protocol, client code can declare types conforming to this protocol. The compiler generates a table of function pointers known as the protocol witness table to describe each protocol conformance. Calling a protocol requirement on a generic parameter requires loading the right function pointer from a protocol witness table. Since protocol requirements can be re-ordered, and new protocol requirements with default implementations can be added, the layout of a protocol witness table must be completely opaque across a resilience boundary.

> This is accomplished in two steps. First, for every protocol requirement, the binary framework exports a special function called a dispatch thunk. The dispatch thunk is part of the framework itself, and so it can directly hard-code the offset of the protocol requirement in the witness table. If the protocol’s declaration is changed to re-order requirements, the order of entries in the witness table is changed, but the symbol names of the dispatch thunks remain the same. Since client code calls all protocol methods via dispatch thunks, binary compatibility with future versions of the framework can be maintained.

> Finally, to cope with adding new protocol requirements, protocol witness tables require runtime instantiation. Instead of emitting a witness table in the client code directly, the compiler emits a symbolic description of the conformance. The instantiation process places the protocol requirements in the correct order and fills missing entries to point to their default implementation, to produce a well-formed witness table which can be passed off to a dispatch thunk.

Reply:
In this case, we should be able to emit the symbolic description of the conformance and the dispatch thunk as Swift expects to call them. Then, because the Swift code is compiled with `enable-library-evolution`, the Swift compiler will emit the code to call the protocol implementations through the dispatch thunks.